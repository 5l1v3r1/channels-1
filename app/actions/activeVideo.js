// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
