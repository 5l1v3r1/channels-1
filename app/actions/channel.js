// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

export function selectChannel(id: string): Action {
  return {
    type: actionTypes.SELECT_CHANNEL,
    payload: id
  };
}
