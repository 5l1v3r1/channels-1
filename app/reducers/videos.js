// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action, VideosState } from '../constants/typeAliases';
import VideoType from '../dataTypes/videoType';
import { stringOrEmpty } from '../utils/utils';

function byId(state: { [string]: VideoType } = {}, { type, payload }: Action): { [string]: VideoType } {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      const nextState = { ...state };
      payload.items.forEach(item => {
        nextState[item.id] = VideoType.from(item);
      });
      return nextState;
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return {};
    default:
      return state;
  }
}

function allIds(state: string[] = [], { type, payload }: Action): string[] {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      return [
        ...state,
        ...payload.items.map(item => stringOrEmpty(item.id))
      ];
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return [];
    default:
      return state;
  }
}

function pageToken(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      return payload.nextPageToken;
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return '';
    default:
      return state;
  }
}

function isFetching(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_REQUEST:
      return true;
    case actionTypes.FETCH_VIDEOS_SUCCESS:
    case actionTypes.FETCH_VIDEOS_FAILURE:
      return false;
    default:
      return state;
  }
}

function error(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_REQUEST:
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      return false;
    case actionTypes.FETCH_VIDEOS_FAILURE:
      return true;
    default:
      return state;
  }
}

const videos = combineReducers({
  byId,
  allIds,
  pageToken,
  isFetching,
  error
});

export function getVideos(state: VideosState): VideoType[] {
  return state.allIds.map(id => state.byId[id]);
}

export default videos;
