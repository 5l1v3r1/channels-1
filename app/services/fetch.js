// @flow
export const Methods = {
  GET: 'GET',
  POST: 'POST'
};

export function authRequest(uri: string): Promise<any> {
  const [url, body] = uri.split('?');
  const init = {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body
  };
  return fetch(url, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function getRequest(uri: string): Promise<any> {
  const init = {
    method: Methods.GET,
    headers: { 'Content-Type': 'application/json' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function postRequest(uri: string, body?: {}): Promise<any> {
  const init = body ?
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  } :
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res;
    });
}
