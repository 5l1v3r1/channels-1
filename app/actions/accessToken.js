// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json.access_token)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
  };
}

export function receiveAccessToken(token: string): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
    payload: token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
  };
}
