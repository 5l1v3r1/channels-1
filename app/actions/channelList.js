// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
