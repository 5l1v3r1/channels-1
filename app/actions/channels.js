// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { token } = getState().accessToken;
    const subscriptions = fetchSubscriptions(token);
    return subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(token, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          return dispatch(receiveChannels(json.items));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

export function fetchSubscriptions(token: string): Promise<any> {
  const uri = getSubscriptionsUri(token);
  return apiRequest(uri, Methods.GET);
}

export function requestChannels(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_REQUEST
  };
}

export function receiveChannels(items: any): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_SUCCESS,
    payload: items
  };
}

export function channelsError(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_FAILURE
  };
}

export function updateActiveChannel(channelId: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { activeId } = getState().channels;
    if (activeId !== channelId) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_CHANNEL,
        payload: channelId
      });
    }
  };
}
