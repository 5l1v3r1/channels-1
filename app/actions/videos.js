// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/types';
import { IVideo } from '../dataTypes/videoType';
import { getRequest } from '../services/fetch';
import { getVideosUri, getVideoIdsUri } from '../services/uriGenerator';

export function fetchVideos(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchVideosRequest());
    const params = {
      channelId: getState().channels.activeId,
      pageToken: getState().videos.pageToken,
      order: 'date'
    };
    return fetchVideoIds(params)
      .then(
        res => {
          const { videoIds, nextPageToken } = res;
          const uri = getVideosUri(videoIds);
          return getRequest(uri)
            .then(
              json => dispatch(fetchVideosSuccess(json.items, nextPageToken)),
              () => dispatch(fetchVideosFailure())
            );
        },
        () => dispatch(fetchVideosFailure())
      );
  };
}

export function fetchVideoIds(params: any): Promise<any> {
  const uri = getVideoIdsUri(params);
  return getRequest(uri)
    .then(
      json => Promise.resolve({
        videoIds: json.items.map(item => item.id.videoId).join(),
        nextPageToken: json.nextPageToken || params.pageToken
      }),
      () => Promise.reject()
    );
}

export function fetchVideosRequest(): Action {
  return {
    type: actionTypes.FETCH_VIDEOS_REQUEST
  };
}

export function fetchVideosSuccess(items: IVideo[], nextPageToken: string): Action {
  return {
    type: actionTypes.FETCH_VIDEOS_SUCCESS,
    payload: {
      items,
      nextPageToken
    }
  };
}

export function fetchVideosFailure(): Action {
  return {
    type: actionTypes.FETCH_VIDEOS_FAILURE
  };
}
