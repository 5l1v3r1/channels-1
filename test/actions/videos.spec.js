/* eslint promise/always-return: 0 */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as videosActions from '../../app/actions/videos';
import * as fetch from '../../app/services/fetch';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../../app/services/uriGenerator');

describe('video list actions', () => {

  const accessToken = 'ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg';
  const channels = {
    activeId: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
  };
  const videos = {
    pageToken: ''
  };

  it('creates FETCH_VIDEOS_SUCCESS when fetching video ids has resolved with undefined', () => {
    const store = mockStore({ accessToken, channels, videos });

    fetch.apiRequest = jest.fn(() => Promise.resolve({}));

    return store.dispatch(videosActions.fetchVideos()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_VIDEOS_SUCCESS when fetching video ids has resolved with null', () => {
    const store = mockStore({ accessToken, channels, videos });

    fetch.apiRequest = jest.fn(() => Promise.resolve({ items: null }));

    return store.dispatch(videosActions.fetchVideos()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_VIDEOS_FAILURE when fetching video ids has been rejected', () => {
    const store = mockStore({ accessToken, channels, videos });

    fetch.apiRequest = jest.fn(() => Promise.reject());

    return store.dispatch(videosActions.fetchVideos()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_VIDEOS_SUCCESS when fetching videos has been resolved', () => {
    const store = mockStore({ accessToken, channels, videos });

    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            id: {
              videoId: 'lAJWHHUz8_8'
            }
          },
          {
            id: {
              videoId: '6YBV1cKRqzU'
            }
          }
        ],
        nextPageToken: 'CBQQAA'
      }))
      .mockReturnValueOnce(Promise.resolve({ items: ['GraphQL Basics - Fun Fun Function', 'Advanced Dependency Injection without classes - Fun Fun Function'] }));

    return store.dispatch(videosActions.fetchVideos()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_VIDEOS_FAILURE when fetching videos has been rejected', () => {
    const store = mockStore({ accessToken, channels, videos });

    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            id: {
              videoId: 'lAJWHHUz8_8'
            }
          },
          {
            id: {
              videoId: '6YBV1cKRqzU'
            }
          }
        ],
        nextPageToken: 'CBQQAA'
      }))
      .mockReturnValueOnce(Promise.reject());

    return store.dispatch(videosActions.fetchVideos()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('should create an action to request videos', () => {
    expect(videosActions.requestVideos()).toMatchSnapshot();
  });

  it('should create an action to receive videos', () => {
    const items = ['GraphQL Basics - Fun Fun Function', 'Advanced Dependency Injection without classes - Fun Fun Function'];
    const nextPageToken = 'CBQQAA';
    expect(videosActions.receiveVideos(items, nextPageToken)).toMatchSnapshot();
  });

  it('should create an action to handle an error', () => {
    expect(videosActions.videosError()).toMatchSnapshot();
  });
});
