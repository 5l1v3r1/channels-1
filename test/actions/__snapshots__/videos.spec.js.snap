// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`video list actions creates FETCH_VIDEOS_FAILURE when fetching video ids has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_VIDEOS_REQUEST",
  },
  Object {
    "type": "FETCH_VIDEOS_FAILURE",
  },
]
`;

exports[`video list actions creates FETCH_VIDEOS_FAILURE when fetching videos has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_VIDEOS_REQUEST",
  },
  Object {
    "type": "FETCH_VIDEOS_FAILURE",
  },
]
`;

exports[`video list actions creates FETCH_VIDEOS_SUCCESS when fetching video ids has resolved with null 1`] = `
Array [
  Object {
    "type": "FETCH_VIDEOS_REQUEST",
  },
  Object {
    "payload": Object {
      "items": Array [],
      "nextPageToken": "",
    },
    "type": "FETCH_VIDEOS_SUCCESS",
  },
]
`;

exports[`video list actions creates FETCH_VIDEOS_SUCCESS when fetching video ids has resolved with undefined 1`] = `
Array [
  Object {
    "type": "FETCH_VIDEOS_REQUEST",
  },
  Object {
    "payload": Object {
      "items": Array [],
      "nextPageToken": "",
    },
    "type": "FETCH_VIDEOS_SUCCESS",
  },
]
`;

exports[`video list actions creates FETCH_VIDEOS_SUCCESS when fetching videos has been resolved 1`] = `
Array [
  Object {
    "type": "FETCH_VIDEOS_REQUEST",
  },
  Object {
    "payload": Object {
      "items": Array [
        "GraphQL Basics - Fun Fun Function",
        "Advanced Dependency Injection without classes - Fun Fun Function",
      ],
      "nextPageToken": "CBQQAA",
    },
    "type": "FETCH_VIDEOS_SUCCESS",
  },
]
`;

exports[`video list actions should create an action to handle an error 1`] = `
Object {
  "type": "FETCH_VIDEOS_FAILURE",
}
`;

exports[`video list actions should create an action to receive videos 1`] = `
Object {
  "payload": Object {
    "items": Array [
      "GraphQL Basics - Fun Fun Function",
      "Advanced Dependency Injection without classes - Fun Fun Function",
    ],
    "nextPageToken": "CBQQAA",
  },
  "type": "FETCH_VIDEOS_SUCCESS",
}
`;

exports[`video list actions should create an action to request videos 1`] = `
Object {
  "type": "FETCH_VIDEOS_REQUEST",
}
`;
